docs/proposals/port_detection_optimizations.md

```diff
@@ -0,0 +1,93 @@
+| Port detection not implemented yet                  | [dotnet](https://github.com/redhat-developer/alizer/blob/main/go/pkg/apis/enricher/framework/dotnet/dotnet_detector.go#L54)    
+
+*As a result the golang related frameworks are using all component paths in order to apply port detection.*
+
+### GetPortDetectionPaths
+A way to ensure consistency on doing port detection, could be to define `GetPortDetectionPaths` inside every detector. This function could provide a list of files to the `DoPortDetection`.
+
+```golang
+// It returns a list of accepted paths or specific files from a given rootPath
+func (e BeegoDetector) GetPortDetectionFiles(rootPath string) []string {
+	paths := []string{"app.conf"}
+	return utils.ReadComponentFiles(rootPath, paths)
+}
+
+func ReadComponentFiles(rootPath string, paths []string) []string {
+// This function should merge the logic of utils.readAnyApplicationFile()
+}
+```
+
+*After this update, as we will already have read all files, we will have to update the go_detector.GetPortFromFilesGo in order not to perform the same process again.*
+
+As a result all detectors will use the same process to read specific files for port detection.
+
+### Process Overview
+The final process for each detector will be:
+
+## Implement Combined and No Strategy Options
+Currently, upon `component` command we have the `port-detection` arg which specifies the port-detection strategy
+
+### Current State
+On the current state as we allow only three values for this arg:
```

Comment:
1. Is there any benefit to requiring the user to specify this arg? During component detection, we already know the type of component we have detected. Should we just use that detected info to decide which type of port detection we are using for each detected component?

2. We need to document the default value if that arg is not specified.

Reply:
re1: Pretty much this is the idea behind `GetPortDetectionPaths` (mentioned above). This function will ensure each time we fetch only the necessary files for port detection. This section offers an additional option of no-port-detection in case the user is not interested.

re2: The default behavior/strategy is mentioned here (https://github.com/devfile/alizer/blob/main/docs/public/port_detection.md#port-detection-in-alizer), but I think too it needs an update

Reply:
1: It is still not clear that we need that args. We have the option to decide if port detection is needed. If that is false, then we'll skip the port detection logic completely. If port detection is true (I think that is the default), we'll go through the port detection logic. 

IIRC...We always do component detection before port detection. When we do component detection, we already know the type from the component detection. Therefore, do we still need the user to explicitly specify the port detection type given that we can get that info from the result of the component detection?

Reply:
> It is still not clear that we need that args. We have the option to decide if port detection is needed. If that is false, then we'll skip the port detection logic completely. If port detection is true (I think that is the default), we'll go through the port detection logic.

Right now, we already have the `port-detection` arg which has been introduced long time ago. The accepted values are `docker`, `compose` and `source`. The default value is `[docker, compose, source]`.

I agree that we can remove the `port-detection` arg. But, It will be difficult to remove this arg from the CLI as it will be non backwards-compatible change. Maybe we can only introduce a `no-port-detection` arg which will skip the port detection flow.

Reply:
> What is the current behaviour if the user doesn't specify port-detection flag today
The default behavior is `docker`, `compose`, `source`:
1. First we detect components. After component detection we are trying to "enrich" them by detecting frameworks and ports.
2. For each component we are trying to find a `dockerfile`/`containerfile` on the root level or one level below.
3. If no success on (2) we do the same for `compose`.
4. If no success on (3) we go for the source code analysis.

> With the current design, the detected port number can use a different strategy than the detected component type and causes a mismatch if the user doesn't specify the port detection policy. That's where we want to use the component type to decide the port number detection strategy by default.

I'm not sure if I understand this comment. Each component is assigned to an enricher according to its language. For example a go component will be assigned to the go enricher. This enricher will follow the given strategy (e.g. `docker`, `compose`, `source`) and will apply the detection flow accordingly.

Reply:
I've updated the proposal. We will focus only on the `port-detection` part as discussed.

I have updated the epic https://github.com/devfile/api/issues/1154 for component-detection optimazations.